### PR TITLE
[FIX] google_spreadsheet: migrate to new api (v3 -> v4)

### PR DESCRIPTION
**Pull request by nboulif**
Steps to Reproduce:

  - Install 'account' module
  - Go to Settings and set a token for Google Drive
  - Activate Google Spreadsheet
  - Go to Invoicing -> Invoices
  - Under Search bar, click on Favorites -> Add to Google Spreadsheet

Issue:

  A spreadsheet does open but with no data (no formula or config).

Cause:

  The V3 API we were using was turned down in August 2021 and therefore
  not able to use the api to write on the spreadsheet.
  https://cloud.google.com/blog/products/g-suite/migrate-your-apps-use-latest-sheets-api)

Note: In case that the system parameters `​google_drive_client_id` and
`google_drive_client_secret` has been changed (and therefore also the
default templates), it must be ensured that the `client` has the
Google Sheet API (v4) activated and the scope
`[https://www.googleapis.com/auth/spreadsheets`](https://www.googleapis.com/auth/spreadsheets%60) set.

API activation and configuration available here :
https://console.developers.google.com/apis/dashboard?project=[PROJECT]
https://console.developers.google.com/apis/credentials/consent/edit?project=[PROJECT]
**[PROJECT]** : Google project that will be linked to `​google_drive_client_id`.

opw-[2633951](https://www.odoo.com/web#id=2633951&view_type=form&model=project.task)

### CODE REVIEW

**Review by nle-odoo — APPROVED** [review]
looks good to me (and the code is easier to read)

this does not say why this is done, from what I understand:

> Issue: when creating a spreadsheet with a formula (eg. when searching res.partner) the formula was not set and the first page contained no data
> Cause: the V3 API we were using was turned down in August 2021 (https://cloud.google.com/blog/products/g-suite/migrate-your-apps-use-latest-sheets-api)

**Comment by nle-odoo** [comment]
current commit looks good to me, from my tests if I had a refresh token already set:

- before the fix: the sheet is created without page customized
- after the fix: the sheet is created with page customized

which make sense since the scope from odoo (get_google_scope) is changed but from the v3 migration documentation they are alias:

> https://developers.google.com/sheets/api/guides/migration
> The Sheets API v3 operates with a single authorization scope: `https://spreadsheets.google.com/feeds` which is an alias for `https://www.googleapis.com/auth/spreadsheets`

For the case of a customer that would have changed the google_drive_client_id/google_drive_client_secret to one that doesn't allow the API `Google Sheet API (v4)` with right scope:

- if his code did not work with API v3 => the change looks ok to me
- if his code worked with API v3 (not sure how, maybe they grandfathered people that had token before the shutdown) => something could be done (keeping both code and fallbacking on v4?) but I'm not sure if this is possible (for the ticket reporting the issue it's not the case at least)

So for me ok to merge, but it would be better to have second review.

**Comment by nle-odoo** [comment]
robodoo delegate+

bon checked with dle and safe to me

as I said, the only issue I could see compared to now, would be it was still working for customer with the old API, but as far as I can tell this is not the case, if it was revealed by reports to be the case it would be simple to fix the use case in stable (=> fallback on V3 API if V4 doesn't work)

**Comment by nboulif** [comment]
@nle-odoo Thanks Again!
robodoo r+

**Pull request closed without merging**
